controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,848 @@
+		changeLeader:     make(chan bool),
+		backoffPeriod:    getBackoffPeriod(),
+		log:              log15.New("component", "scheduler"),
+		hostStreams:      make(map[string]stream.Stream),
+		jobs:             make(map[string]*Job),
+		pendingStarts:    make(pendingJobs),
+		pendingStops:     make(pendingJobs),
+		formations:       make(Formations),
+		hostEvents:       make(chan *host.Event, eventBufferSize),
+		listeners:        make(map[chan Event]struct{}),
+		stop:             make(chan interface{}),
+		rectifyJobs:      make(chan interface{}, 1),
+		syncJobs:         make(chan interface{}, 1),
+		syncFormations:   make(chan interface{}, 1),
+		formationChange:  make(chan *ct.ExpandedFormation, eventBufferSize),
+		hostChange:       make(chan *discoverd.Event, eventBufferSize),
+		jobRequests:      make(chan *JobRequest, eventBufferSize),
+		putJobs:          make(chan *ct.Job, eventBufferSize),
+	}
+}
+
+func main() {
+	clusterClient := utils.ClusterClientWrapper(cluster.NewClient())
+	controllerClient, err := controller.NewClient("", os.Getenv("AUTH_KEY"))
+	if err != nil {
+		shutdown.Fatal(err)
+	}
+	s := NewScheduler(clusterClient, controllerClient)
+	if err := s.Run(); err != nil {
+		shutdown.Fatal(err)
+	}
+	shutdown.Exit()
+}
+
+func (s *Scheduler) Run() error {
+	log := s.log.New("fn", "Run")
+	log.Info("starting scheduler loop")
+	defer log.Info("exiting scheduler loop")
+
+	tickChannel(s.syncJobs, 30*time.Second)
+	tickChannel(s.syncFormations, time.Minute)
+
+	stream, err := s.StreamHostEvents(s.hostChange)
+	if err != nil {
+		return fmt.Errorf("Unable to stream hosts. Error: %v", err)
+	}
+	defer stream.Close()
+
+	stream, err = s.StreamFormations(nil, s.formationChange)
+	if err != nil {
+		return fmt.Errorf("Unable to stream formations. Error: %v", err)
+	}
+	defer stream.Close()
+
+	go s.RunPutJobs()
+
+	err = nil
+	for {
+		if err != nil {
+			log.Error("An error occurred", "error", err)
+			err = nil
+		}
+		select {
+		case <-s.stop:
+			close(s.putJobs)
+			return nil
+		case isLeader := <-s.changeLeader:
+			s.HandleLeaderChange(isLeader)
+			continue
+		default:
+		}
+
+		// Handle events that reconcile scheduler state with the cluster
+		select {
+		case req := <-s.jobRequests:
+			err = s.HandleJobRequest(req)
+			continue
+		case e := <-s.hostChange:
+			err = s.HandleHostChange(e)
+			continue
+		case he := <-s.hostEvents:
+			err = s.HandleHostEvent(he)
+			continue
+		case ef := <-s.formationChange:
+			err = s.FormationChange(ef)
+			continue
+		default:
+		}
+
+		// Handle sync events
+		select {
+		case <-s.syncFormations:
+			err = s.SyncFormations()
+			continue
+		case <-s.syncJobs:
+			err = s.SyncJobs()
+			continue
+		default:
+		}
+
+		// Finally, handle triggering cluster changes
+		select {
+		case <-s.rectifyJobs:
+			err = s.RectifyJobs()
+		case <-time.After(50 * time.Millisecond):
+			// block so that
+			//	1) mutate events are given a chance to happen
+			//  2) we don't spin hot
```

Comment:
It probably makes sense to copy all of the unique channel receives from above into this select so that we don't need a timer.